### PR TITLE
[PM-7290] replace legacy abstraction with generation algorithms

### PR DESCRIPTION
**Pull request by audrey-jensen**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-7290

## 📔 Objective

Remove MV2 services

## 🧾 Changes

* Move generation algorithms into strategies
* Introduce Randomizer
* Introduce Policy configurations
* Reduce repetition in strategies
* Remove MV2 generators - algorithms integrated with strategies

## 🦕 System evolution

Strategies are likely to be decomposed into generator configurations and engines in a future PR. The configurations load and store settings, policy controls, and lookup metadata. The engines apply the policy and generate credentials. This design allows the construction of rx operators and angular services from the decomposed parts.

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: libs/common/src/tools/generator/username/forwarders/firefox-relay.ts [hunk 1]
```diff
@@ -16,6 +13,7 @@ export const DefaultFirefoxRelayOptions: ApiOptions = Object.freeze({
   token: "",
 });
 
+/** Generates a forwarding address for Firefox Relay */
```
Comment (djsmith85) [inline libs/common/src/tools/generator/username/forwarders/firefox-relay.ts:16]: ⛏️  Duplicate line

File: libs/common/src/tools/generator/username/forwarders/duck-duck-go.ts [hunk 2]
```diff
@@ -16,7 +13,6 @@ export const DefaultDuckDuckGoOptions: ApiOptions = Object.freeze({
   token: "",
 });
 
-/** Generates a forwarding address for DuckDuckGo */
```
Comment (djsmith85) [inline libs/common/src/tools/generator/username/forwarders/duck-duck-go.ts:19]: ❓ Why are you removing this doc-line?

**Review by djsmith85 — APPROVED** [review]
@audrey-jensen Awesome work! Great idea to split the previous PR up into more consumable/reviewable parts.

Really only some minor nits (docs updates and questions). I'm happy to approve now, as their only nits and I know you are working on follow-ups and taking my feedback into consideration for the upcoming PRs (especially documentation 😁 ).

Regardless, let me know if you'd still address them here and I'll re-review.

File: libs/common/src/tools/generator/username/subaddress-generator-strategy.ts [hunk 3]
```diff
@@ -26,34 +24,42 @@ export class SubaddressGeneratorStrategy
...
+  durableState = sharedStateByUserId(SUBADDRESS_SETTINGS, this.stateProvider);
+  defaults$ = clone$PerUserId(this.defaultOptions);
+  toEvaluator = newDefaultEvaluator<SubaddressGenerationOptions>();
+  readonly policy = PolicyType.PasswordGenerator;
 
-  /** {@link GeneratorStrategy.defaults$} */
-  defaults$(userId: UserId) {
-    return new BehaviorSubject({ ...DefaultSubaddressOptions }).asObservable();
-  }
+  // algorithm
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/subaddress-generator-strategy.ts:38]: It's documented by the [GeneratorStrategy](https://github.com/bitwarden/clients/blob/939c20d1873b3996e0c788f9863e1c591e6c5fe4/libs/common/src/tools/generator/abstractions/generator-strategy.abstraction.ts#L13) abstraction.

File: libs/common/src/tools/generator/password/password-generator-strategy.ts [hunk 4]
```diff
@@ -27,36 +21,108 @@ export class PasswordGeneratorStrategy
...
-  /** {@link GeneratorStrategy.durableState} */
-  durableState(id: UserId) {
-    return this.stateProvider.getUser(id, PASSWORD_SETTINGS);
+  // configuration
+  durableState = sharedStateByUserId(PASSWORD_SETTINGS, this.stateProvider);
+  defaults$ = clone$PerUserId(DefaultPasswordGenerationOptions);
+  readonly policy = PolicyType.PasswordGenerator;
+  toEvaluator() {
+    return mapPolicyToEvaluator(Policies.Password);
   }
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/password/password-generator-strategy.ts:34]: The default evaluator is a no-op evaluator. The password and passphrase strategies construct their evaluators using `Policies.Password` and `Policies.Passphrase` respectively.

File: libs/common/src/tools/generator/username/forwarders/duck-duck-go.ts (hunk 2 above)
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/forwarders/duck-duck-go.ts:19]: It's documented by the [GeneratorStrategy](https://github.com/bitwarden/clients/blob/939c20d1873b3996e0c788f9863e1c591e6c5fe4/libs/common/src/tools/generator/abstractions/generator-strategy.abstraction.ts#L13) abstraction.

File: libs/common/src/tools/generator/username/forwarders/duck-duck-go.ts [hunk 5]
```diff
@@ -32,23 +28,12 @@ export class DuckDuckGoForwarder extends ForwarderGeneratorStrategy<ApiOptions>
...
-
-  /** {@link ForwarderGeneratorStrategy.defaults$} */
-  defaults$ = (userId: UserId) => {
-    return new BehaviorSubject({ ...DefaultDuckDuckGoOptions });
-  };
+  // configuration
+  readonly key = DUCK_DUCK_GO_FORWARDER;
+  readonly rolloverKey = DUCK_DUCK_GO_BUFFER;
 
   /** {@link ForwarderGeneratorStrategy.generate} */
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/forwarders/duck-duck-go.ts]: Addressed in 7d8976e.

File: libs/common/src/tools/generator/username/forwarders/fastmail.ts [hunk 6]
```diff
@@ -34,23 +31,12 @@ export class FastmailForwarder extends ForwarderGeneratorStrategy<ApiOptions & E
...
-
-  /** {@link ForwarderGeneratorStrategy.rolloverKey} */
-  get rolloverKey() {
-    return FASTMAIL_BUFFER;
-  }
+  // configuration
+  readonly key = FASTMAIL_FORWARDER;
+  readonly rolloverKey = FASTMAIL_BUFFER;
 
   /** {@link ForwarderGeneratorStrategy.generate} */
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/forwarders/fastmail.ts]: Addressed in 7d8976e.

File: libs/common/src/tools/generator/username/forwarders/firefox-relay.ts [hunk 7]
```diff
@@ -32,23 +30,12 @@ export class FirefoxRelayForwarder extends ForwarderGeneratorStrategy<ApiOptions
...
-
-  /** {@link ForwarderGeneratorStrategy.defaults$} */
-  defaults$ = (userId: UserId) => {
-    return new BehaviorSubject({ ...DefaultFirefoxRelayOptions });
-  };
+  // configuration
+  readonly key = FIREFOX_RELAY_FORWARDER;
+  readonly rolloverKey = FIREFOX_RELAY_BUFFER;
 
   /** {@link ForwarderGeneratorStrategy.generate} */
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/forwarders/firefox-relay.ts]: Addressed in 7d8976e.

File: libs/common/src/tools/generator/username/forwarders/forward-email.ts [hunk 8]
```diff
@@ -36,23 +33,12 @@ export class ForwardEmailForwarder extends ForwarderGeneratorStrategy<
...
-
-  /** {@link ForwarderGeneratorStrategy.rolloverKey} */
-  get rolloverKey() {
-    return FORWARD_EMAIL_BUFFER;
-  }
+  // configuration
+  readonly key = FORWARD_EMAIL_FORWARDER;
+  readonly rolloverKey = FORWARD_EMAIL_BUFFER;
 
   /** {@link ForwarderGeneratorStrategy.generate} */
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/forwarders/forward-email.ts]: Addressed in 7d8976e.

File: libs/common/src/tools/generator/password/password-generator-strategy.ts [hunk 9]
```diff
@@ -27,36 +21,108 @@ export class PasswordGeneratorStrategy
...
 
-  /** Gets the default options. */
-  defaults$(_: UserId) {
-    return new BehaviorSubject({ ...DefaultPasswordGenerationOptions }).asObservable();
-  }
+  /** {@link GeneratorStrategy.generate} */
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/password/password-generator-strategy.ts]: Addressed in 7d8976e.

File: libs/common/src/tools/generator/passphrase/passphrase-generator-strategy.ts [hunk 10]
```diff
@@ -28,36 +23,48 @@ export class PassphraseGeneratorStrategy
...
+  readonly policy = PolicyType.PasswordGenerator;
+  toEvaluator() {
+    return mapPolicyToEvaluator(Policies.Passphrase);
   }
 
-  /** Gets the default options. */
-  defaults$(_: UserId) {
-    return new BehaviorSubject({ ...DefaultPassphraseGenerationOptions }).asObservable();
-  }
+  /** {@link GeneratorStrategy.generate} */
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/passphrase/passphrase-generator-strategy.ts]: Addressed in 7d8976e.

File: libs/common/src/tools/generator/username/catchall-generator-strategy.ts [hunk 11]
```diff
@@ -19,34 +17,33 @@ export class CatchallGeneratorStrategy
...
-  /** {@link GeneratorStrategy.toEvaluator} */
-  toEvaluator() {
-    return pipe(map((_) => new DefaultPolicyEvaluator<CatchallGenerationOptions>()));
-  }
+  durableState = sharedStateByUserId(CATCHALL_SETTINGS, this.stateProvider);
+  defaults$ = clone$PerUserId(this.defaultOptions);
+  toEvaluator = newDefaultEvaluator<CatchallGenerationOptions>();
+  readonly policy = PolicyType.PasswordGenerator;
 
   /** {@link GeneratorStrategy.generate} */
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/catchall-generator-strategy.ts]: Addressed in 7d8976e.

File: libs/common/src/tools/generator/random.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
+import { CryptoService } from "@bitwarden/common/platform/abstractions/crypto.service";
+
+import { Randomizer } from "./abstractions/randomizer";
+import { WordOptions } from "./word-options";
+
+export class CryptoServiceRandomizer implements Randomizer {
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/random.ts:7]: The methods are documented by [Randomizer](https://github.com/bitwarden/clients/blob/a4369b005e9b8cbb3239130893320c839a9ddcb9/libs/common/src/tools/generator/abstractions/randomizer.ts).

Documented the class in 7d8976e8

File: libs/common/src/tools/generator/username/forwarders/firefox-relay.ts (hunk 1 above)
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/forwarders/firefox-relay.ts]: Addressed in 7d8976e.

**Review by ttalty — APPROVED** [review]

**Pull request merged**